File: mobile/src/main/java/at/shockbytes/remote/network/message/JsonMessageDeserializer.java
package at.shockbytes.remote.network.message;

import android.util.Base64;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import at.shockbytes.remote.network.model.ConnectionConfig;
import at.shockbytes.remote.network.model.FileTransferResponse;
import at.shockbytes.remote.network.model.KeyExchangeResponse;
import at.shockbytes.remote.network.model.RemiFile;
import at.shockbytes.remote.network.model.SlidesResponse;

/**
 * @author Martin Macheiner
 *         Date: 28.09.2017.
 */

public class JsonMessageDeserializer implements MessageDeserializer {

    private final Gson gson;

    public JsonMessageDeserializer() {
        gson = new Gson();
    }

    @Override
    public List<String> requestAppsMessage(String msg) {
        return gson.fromJson(msg, new TypeToken<ArrayList<String>>() {
        }.getType());
    }

    @Override
    public List<RemiFile> requestFilesMessage(String msg) {
        return gson.fromJson(msg, new TypeToken<ArrayList<RemiFile>>() {
        }.getType());
    }

    @Override
    public ConnectionConfig welcomeMessage(String msg) {

        JsonObject element = new JsonParser().parse(msg).getAsJsonObject();

        JsonObject permObj = element.get("permissions").getAsJsonObject();
        ConnectionConfig.ConnectionPermissions permissions =
                new ConnectionConfig.ConnectionPermissions(permObj.get("perm_mouse").getAsBoolean(),
                        permObj.get("perm_files").getAsBoolean(),
                        permObj.get("perm_file_transfer").getAsBoolean(),
                        permObj.get("perm_apps").getAsBoolean());
        return new ConnectionConfig(element.get("operating_system").getAsString(), permissions);
    }

    @Override
    public FileTransferResponse fileTransferMessage(String msg) {

        try {

            JSONObject object = new JSONObject(msg);

            String filename = object.getString("filename");
            long size = object.getLong("size");
            int transferCode = object.getInt("transferCode");
            byte[] content = new byte[0];
            String exception = "";
            if (size > 0) {
                content = Base64.decode(object.getString("content"), Base64.DEFAULT);
            } else {
                exception = object.getString("exception");
            }
            return new FileTransferResponse(filename, content, transferCode, exception);

        } catch (JSONException e) {
            e.printStackTrace();
        }
        return new FileTransferResponse();
    }

    @Override
    public SlidesResponse requestSlides(String msg) {
        return gson.fromJson(msg, SlidesResponse.class);
    }

    @Override
    public KeyExchangeResponse keyExchangeResponse(String msg) {
        JsonObject element = new JsonParser().parse(msg).getAsJsonObject();
        return new KeyExchangeResponse(element.get("desktop").getAsString(),
                element.get("certificate").getAsString());
    }
}

File: mobile/src/main/java/at/shockbytes/remote/network/message/JsonMessageDeserializer.kt
package at.shockbytes.remote.network.message

import android.util.Base64
import at.shockbytes.remote.network.model.*
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import java.util.*

/**
 * @author Martin Macheiner
 * Date: 28.09.2017.
 */

class JsonMessageDeserializer : MessageDeserializer {

    private val gson: Gson = Gson()

    override fun requestAppsMessage(msg: String): List<String> {
        return gson.fromJson<List<String>>(msg, object : TypeToken<ArrayList<String>>() {}.type)
    }

    override fun requestFilesMessage(msg: String): List<RemiFile> {
        return gson.fromJson<List<RemiFile>>(msg, object : TypeToken<ArrayList<RemiFile>>() {}.type)
    }

    override fun welcomeMessage(msg: String): ConnectionConfig {

        val element = JsonParser().parse(msg).asJsonObject
        val permissionObject = element.get("permissions").asJsonObject

        val permissions = ConnectionConfig.ConnectionPermissions(
                permissionObject.get("perm_mouse").asBoolean,
                permissionObject.get("perm_files").asBoolean,
                permissionObject.get("perm_file_transfer").asBoolean,
                permissionObject.get("perm_apps").asBoolean)
        return ConnectionConfig(element.get("operating_system").asString, permissions)
    }

    override fun fileTransferMessage(msg: String): FileTransferResponse {

        val element = JsonParser().parse(msg).asJsonObject

        val filename = element.get("filename").asString
        val size = element.get("size").asLong
        val transferCode = element.get("transferCode").asInt
        val exception: String? = element.get("exception")?.asString

        var content: ByteArray? = null
        if (size > 0) {
            content = Base64.decode(element.get("content").asString, Base64.DEFAULT)
        }
        return FileTransferResponse(filename, transferCode, content, exception)
    }

    override fun requestSlides(msg: String): SlidesResponse {
        return gson.fromJson(msg, SlidesResponse::class.java)
    }

    override fun keyExchangeResponse(msg: String): KeyExchangeResponse {
        val element = JsonParser().parse(msg).asJsonObject
        return KeyExchangeResponse(element.get("desktop").asString,
                element.get("certificate").asString)
    }
}
